src/shared/openssl-util.c

```diff
@@ -573,6 +573,215 @@ int string_hashsum(
 
 }
 #  endif
+
+int ecdh_derive_shared_secret(
+                EVP_PKEY *private_key,
+                EVP_PKEY *public_key,
+                void **ret_shared_secret,
+                size_t *ret_shared_secret_size) {
+
+        _cleanup_(EVP_PKEY_CTX_freep) EVP_PKEY_CTX *ctx = NULL;
+        _cleanup_(erase_and_freep) void *shared_secret = NULL;
+        size_t shared_secret_size;
+        int r;
+
+        assert(private_key);
+        assert(public_key);
+        assert(ret_shared_secret);
+        assert(ret_shared_secret_size);
+
+        ctx = EVP_PKEY_CTX_new(private_key, NULL);
+        if (!ctx)
+                return log_openssl_errors("Failed to allocate pkey context");
+
+        r = EVP_PKEY_derive_init(ctx);
+        if (r != 1)
+                return log_openssl_errors("Failed to initialize pkey derive context");
+
+        r = EVP_PKEY_derive_set_peer(ctx, public_key);
+        if (r != 1)
+                return log_openssl_errors("Failed to set peer key for derivation");
+
+        r = EVP_PKEY_derive(ctx, NULL, &shared_secret_size);
+        if (r != 1)
+                return log_openssl_errors("Failed to get shared secret size");
+
+        if (shared_secret_size == 0)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to get shared secret size");
+
+        shared_secret = malloc(shared_secret_size);
+        if (!shared_secret)
+                return log_oom();
+
+        r = EVP_PKEY_derive(ctx, shared_secret, &shared_secret_size);
+        if (r != 1)
+                return log_openssl_errors("Failed to derive shared secret");
+
+        *ret_shared_secret = TAKE_PTR(shared_secret);
+        *ret_shared_secret_size = shared_secret_size;
+        return 0;
+}
+
+static int ecc_pkey_generate_volume_key(
+                EVP_PKEY *pkey,
+                void **ret_decrypted_key,
+                size_t *ret_decrypted_key_size,
+                void **ret_saved_key,
+                size_t *ret_saved_key_size) {
+
+        _cleanup_(EVP_PKEY_freep) EVP_PKEY *pkey_new = NULL;
+        _cleanup_(erase_and_freep) void *decrypted_key = NULL;
+        _cleanup_free_ unsigned char *saved_key = NULL;
+        size_t decrypted_key_size, saved_key_size;
+        int nid = NID_undef;
+        int r;
+
+#if OPENSSL_VERSION_MAJOR >= 3
+        _cleanup_free_ char *curve_name = NULL;
+        size_t len = 0;
+        if (EVP_PKEY_get_group_name(pkey, NULL, 0, &len) != 1 || len == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to determine PKEY group name length");
+        len++;
+        curve_name = new(char, len);
+        if (!curve_name)
+                return log_oom();
+
+        if (!EVP_PKEY_get_group_name(pkey, curve_name, len, &len))
```

Comment:
above you check for the error with `!= 1`, here with `!`, i.e. `== 0`. Please stick to one form. And it probably should be `!= 1` since we do that almost everywhere else for our openssl calls